### PR TITLE
fix(android/app,oem/fv/android): Sanitize Sentry navigation breadcrumbs

### LINKED ISSUE #3997: bug(android/app): Sentry breadcrumbs leaking too much info
Seen in Sentry log https://sentry.keyman.com/organizations/keyman/issues/1977

We need to filter breadcrumbs in Sentry for KMW embedded to avoid navigation that reports keystroke events.

References
https://docs.sentry.io/platforms/javascript/enriching-events/breadcrumbs/#customize-breadcrumbs

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #3997

This sanitizes the Sentry navigation breadcrumbs so they don't include key information.

For testing the breadcrumb, I manually created breadcrumbs in `KMManager.shouldOverrideUrlLoading()` for "showKeyPreview"
```
        Breadcrumb breadcrumb = new Breadcrumb();
        breadcrumb.setCategory("navigation");
        breadcrumb.setType("navigation");
        breadcrumb.setData("from", url);
        breadcrumb.setData("to", url);
        breadcrumb.setLevel(SentryLevel.INFO);
        Sentry.addBreadcrumb(breadcrumb);
```

Reference

### CODE REVIEW

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java [hunk 1]
```diff
@@ -129,6 +130,23 @@ protected void onCreate(Bundle savedInstanceState) {
     checkSendCrashReport();
     if (KMManager.getMaySendCrashReport()) {
       SentryAndroid.init(context, options -> {
+        options.setBeforeBreadcrumb((breadcrumb, hint) -> {
+          String NAVIGATION_PATTERN = "^(.*)?(keyboard\\.html#[^-]+)-(.)*$";
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java:134]:
```suggestion
          String NAVIGATION_PATTERN = "^(.*)?(keyboard\\.html#[^-]+)-.*$";
```

It's the same but removes the unnecessary group.

Also, could consider including the digits:

```suggestion
          String NAVIGATION_PATTERN = "^(.*)?(keyboard\\.html#[^-]+-(\d+))(\\+.*)?$";
```

But I'm not sure if there are other patterns we'd need to exclude then?

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by jahorton** [comment]
Do we have any example Sentry reports that verify this fix?

**Comment by darcywong00** [comment]
With the breadcrumb test code above, I halted on [setBeforeBreadrumb](https://github.com/keymanapp/keyman/pull/4025/files#diff-6470252da8d2d4bccb192c79aeedba3607143ad9a4da47da9ca48016413e6792R133) and verified the regex sanitized the from/to fields.

I haven't found a repro for the original issue in Sentry

**Comment by mcdurdin** [comment]
You could just trigger an error using the chrome web inspector -- that should be enough to verify that the breadcrumbs are sanitised

**Comment by jahorton** [comment]
Use of my custom test-keyboard looked good.

Screen Shot 2020-12-02 at 3 47 22 PM

No evidence of the first keystroke.

**Comment by mcdurdin** [comment]
I just tested this against 14.0.194 and I don't believe it is resolved. As Joshua noted, the mitigation was done in Android-side but the breadcrumbs are collected web-side.

I tested with the good old chrome://inspect and `window.setTimeout(() => { throw "error"}, 10);` pattern, which generated the following error:

https://sentry.keyman.com/organizations/keyman/issues/2027/events/cacf8524a976430980ff25f23755b075/

image

**Comment by darcywong00** [comment]
The PR for this was reverted. Since the Navigation breadcrumbs are strictly handled by embedded KMW sentry-manager, this issue is closed (no changes required)

**Pull request merged**